In [80]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [81]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [82]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [83]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [84]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_market_data)

In [85]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(scaled_data, columns=df_market_data.columns, index=df_market_data.index)

# Copy the crypto names from the original DataFrame
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coin_id column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display the scaled DataFrame
df_market_data_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [86]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [87]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=i, random_state=1)

    # 2. Fit the model to the data using `df_market_data_scaled`
    model.fit(df_market_data_scaled)
    
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)

In [88]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k,
    "inertia": inertia
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve = pd.DataFrame(elbow_data)


In [89]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k
df_elbow_curve.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k,
    width=800,
    height=400
)



:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [90]:
# Initialize the K-Means model using the best value for k (4)
model = KMeans(n_clusters=4, random_state=1)


In [91]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_market_data_scaled)


KMeans(n_clusters=4, random_state=1)

In [92]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
scaled_clusters= model.predict(df_market_data_scaled)


# Print the resulting array of cluster values
scaled_clusters

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [93]:
# Create a copy of the scaled DataFrame
df_market_data_scaled_clustered = df_market_data_scaled.copy()

In [94]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_market_data_scaled_clustered["clusters"] = scaled_clusters


# Display the copy of the scaled DataFrame
df_market_data_scaled_clustered.head()



price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters  
coin_id                                             
bitcoin                        -0.251637         2  
ethereum                       -0.199352         2  
tether                         -0.282061         0  
ripple                         -0.295546         0  
bitcoin-cash                   -0.270317         2

In [95]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point
df_market_data_scaled_clustered.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="clusters",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis",
    width=800,
    height=400
)


:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [96]:
# Create a PCA model instance and set `n_components=3`
pca = PCA(n_components=3)


In [97]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components
market_data_pca = pca.fit_transform(df_market_data_scaled)

#View the scaled PCA data
market_data_pca


array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292],
       [-0.24819846, -1.37625159, -1.46202571],
       [-0.43840762, -0.17533654, -0.6633884 ],
       [-0.69342533, -0.47381462, -0.52759693],
       [ 0.06049915,  2.90940385,  1.49857131],
       [-0.39335243, -0.10819197, -0.01275608],
       [-0.79617564, -0.49440875,  1.08281169],
       [ 0.06407452, -1.26982514, -1.09882928],
       [-0.48901506, -0.73271912, -0.06254323],
       [-0.3062723 ,  0.70341515,  1.71422359],
       [-0.51352775, -0.14280239, -0.65656583],
       [-0.36212044, -0.98691441, -0.728

In [98]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component
explained_variance = pca.explained_variance_ratio_

# Print the explained variance
explained_variance



array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.89503166
Total Explained Variance = PC1 + PC2 + PC3
                         = 0.3719856 + 0.34700813 + 0.17603793 = 0.89503166

In [99]:
# Create a new DataFrame with the PCA data
df_market_data_pca = pd.DataFrame(
    market_data_pca,
    columns=["PC1", "PC2", "PC3"]
)
# Copy the crypto names from the original scaled DataFrame
df_market_data_pca["coin_id"] = df_market_data_scaled.index

# Set the coin_id column as index
df_market_data_pca = df_market_data_pca.set_index("coin_id")

# Display the scaled PCA DataFrame
df_market_data_pca.head(5)

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [100]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1, 11))


In [101]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
for i in k_pca:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model_pca = KMeans(n_clusters=i, random_state=1)
    # 2. Fit the model to the data using `df_market_data_pca`
    model_pca.fit(df_market_data_pca)
    # 3. Append the model.inertia_ to the inertia list
    inertia_pca.append(model_pca.inertia_)


In [102]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k": k_pca,
    "inertia": inertia_pca
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve_pca = pd.DataFrame(elbow_data_pca)



In [103]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k
df_elbow_curve_pca.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve with PCA",
    xticks=k_pca,
    width=800,
    height=400
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The k value is the same in both sets of data

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [104]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters=4, random_state=1)

In [105]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=1)

In [106]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
df_market_data_pca["clusters"] = model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values
df_market_data_pca["clusters"]

coin_id
bitcoin                 2
ethereum                2
tether                  0
ripple                  0
bitcoin-cash            2
binancecoin             2
chainlink               2
cardano                 2
litecoin                2
bitcoin-cash-sv         0
crypto-com-chain        0
usd-coin                0
eos                     0
monero                  2
tron                    0
tezos                   2
okb                     0
stellar                 0
cosmos                  2
cdai                    0
neo                     0
wrapped-bitcoin         2
leo-token               0
huobi-token             0
nem                     0
binance-usd             0
iota                    0
vechain                 0
zcash                   2
theta-token             0
dash                    0
ethereum-classic        0
ethlend                 3
maker                   2
havven                  0
omisego                 0
celsius-degree-token    1
ontology                0
ftx-

In [107]:
# Create a copy of the scaled PCA DataFrame
df_market_data_pca_clustered = df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_market_data_pca_clustered["clusters"] = model_pca.labels_

# Display the copy of the scaled PCA DataFrame
df_market_data_pca_clustered.head()


PC1       PC2       PC3  clusters
coin_id                                             
bitcoin      -0.600667  0.842760  0.461595         2
ethereum     -0.458261  0.458466  0.952877         2
tether       -0.433070 -0.168126 -0.641752         0
ripple       -0.471835 -0.222660 -0.479053         0
bitcoin-cash -1.157800  2.041209  1.859715         2

In [109]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point
df_market_data_pca_clustered.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="clusters",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis with PCA",
    width=800,
    height=400
)



:NdOverlay   [clusters]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [110]:
# Composite plot to contrast the Elbow curves
df_elbow_curve.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k,
    width=400,
    height=400
) * df_elbow_curve_pca.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve with PCA",
    xticks=k_pca,
    width=400,
    height=400
)





:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [112]:
# Composite plot to contrast the clusters
df_market_data_scaled_clustered.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="clusters",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis",
    width=400,
    height=400
) + df_market_data_pca_clustered.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="clusters",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis with PCA",
    width=400,
    height=400
)



:Layout
   .NdOverlay.I  :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [clusters]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using PCA for feature reduction in K-Means clustering improves cluster clarity, reduces noise, enhances visualization, and boosts computational efficiency. However, it's crucial to ensure the principal components retain enough variance to accurately represent the original data.